libraries/core_libs/consensus/src/pbft/pbft_manager.cpp

```diff
@@ -1238,26 +1252,65 @@ blk_hash_t PbftManager::proposePbftBlock_() {
   // get DAG block and transaction order
   const auto propose_period = pbft_chain_->getPbftChainSize() + 1;
   auto dag_block_order = dag_mgr_->getDagBlockOrder(dag_block_hash, propose_period);
+
   if (dag_block_order.empty()) {
     LOG(log_er_) << "DAG anchor block hash " << dag_block_hash << " getDagBlockOrder failed in propose";
     assert(false);
   }
-  std::vector<trx_hash_t> non_finalized_transactions;
+
   std::unordered_set<trx_hash_t> trx_hashes_set;
   std::vector<trx_hash_t> trx_hashes;
-  for (auto const &blk_hash : dag_block_order) {
+  u256 total_weight = 0;
+  uint32_t dag_blocks_included = 0;
+  for (auto itr = dag_block_order.begin(); itr != dag_block_order.end(); ++itr) {
+    const auto blk_hash = *itr;
     auto dag_blk = dag_blk_mgr_->getDagBlock(blk_hash);
     if (!dag_blk) {
       LOG(log_er_) << "DAG anchor block hash " << dag_block_hash << " getDagBlock failed in propose for block "
                    << blk_hash;
       assert(false);
     }
+    u256 dag_block_weight = 0;
+    std::vector<trx_hash_t> hashes;
+    const auto &estimations = dag_blk->getEstimations();
+    hashes.reserve(estimations.size());
+
+    int32_t i = 0;
     for (auto const &trx_hash : dag_blk->getTrxs()) {
       if (trx_hashes_set.emplace(trx_hash).second) {
-        trx_hashes.emplace_back(trx_hash);
+        hashes.emplace_back(trx_hash);
+        dag_block_weight += estimations[i];
       }
+      i++;
     }
+    if (total_weight + dag_block_weight > config_.gas_limit) {
+      break;
+    }
+    total_weight += dag_block_weight;
+    dag_blocks_included++;
+  }
+
+  if (dag_blocks_included != dag_block_order.size()) {
+    auto closest_anchor = findClosestAnchor(ghost, dag_block_order, dag_blocks_included);
+    assert(closest_anchor.has_value());
+
+    dag_block_hash = closest_anchor.value();
+    dag_block_order = dag_mgr_->getDagBlockOrder(dag_block_hash, propose_period);
   }
+  {
+    std::unordered_set<trx_hash_t> trx_set;
+    std::vector<trx_hash_t> transactions_to_query;
+    for (auto const &dag_blk_hash : dag_block_order) {
+      auto dag_block = dag_blk_mgr_->getDagBlock(dag_blk_hash);
+      assert(dag_block);
+      for (auto const &trx_hash : dag_block->getTrxs()) {
+        if (trx_set.insert(trx_hash).second) {
+          trx_hashes.emplace_back(trx_hash);
+        }
+      }
+    }
+  }
```

Comment:
can this section be done in upper loop where you calculate size? I mean in case `dag_blocks_included == dag_block_order.size()` then it should be same